lamagraph-core/src/Core/Reducer.hs

```diff
@@ -0,0 +1,199 @@
+{-# LANGUAGE PartialTypeSignatures #-}
+{-# OPTIONS_GHC -Wno-partial-type-signatures #-}
+
+module Core.Reducer where
+
+import Clash.Prelude
+import Core.Node
+
+type NumOfNodesToStore = Unsigned 3
+type NumOfEdgesToStore = Unsigned 3
+type IdOfPort = Unsigned 3
+
+data ReducerStatus
+  = Work
+  | Finished
+  | ErrorHandleSingleNode
+  | ErrorEmptyLeftNode
+  deriving (Generic, NFDataX)
+
+-- | State of the reducer (in terms of mealy automaton).
+data ReducerState
+  = Initial Address
+  | Empty
+  | OneNodeToLoad
+  | Done
+  | Failed
+  deriving (Generic, NFDataX)
+
+-- | Pointer on one of End of `Edge`
+data EdgeEnd = EdgeEnd
+  { _addressOfVertex :: Address
+  , _idOfPort :: IdOfPort
+  }
+  deriving (Generic, NFDataX)
+
+-- | Pair of the ends.
+data Edge = Edge
+  { _leftEnd :: EdgeEnd
+  , _rightEnd :: EdgeEnd
+  }
+  deriving (Generic, NFDataX)
+
+data DataToStore maxNumOfNodesToStore maxNumOfEdgesToStore numberOfPorts = DataToStore
+  { _nodes :: Vec maxNumOfNodesToStore (Maybe (LoadedNode numberOfPorts))
+  , _edges :: Vec maxNumOfEdgesToStore (Maybe Edge)
+  }
+  deriving (Generic, NFDataX)
+
+-- | Select next `Node` to handle and next `Address` by `Port`.
```

Comment:
This comment should be more valuable ad detailed. We select left node and address of right node to be loaded.